File: ludus_mobile_app/lib/features/payment/services/payment_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/payment_model.dart';

class PaymentService {
  static const String baseUrl = 'http://localhost:5000/api'; // Update with your backend URL
  
  // Process payment for a booking
  static Future<PaymentModel> processPayment({
    required String bookingId,
    required String userId,
    required double amount,
    required PaymentMethod method,
    required Map<String, dynamic> paymentDetails,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/payments/process'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: json.encode({
          'bookingId': bookingId,
          'userId': userId,
          'amount': amount,
          'method': method.toString().split('.').last,
          'paymentDetails': paymentDetails,
        }),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body)['data'];
        return PaymentModel.fromJson(data);
      } else {
        throw Exception('Failed to process payment: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error processing payment: $e');
    }
  }

  // Get payment by ID
  static Future<PaymentModel> getPaymentById(String paymentId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/payments/$paymentId'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body)['data'];
        return PaymentModel.fromJson(data);
      } else {
        throw Exception('Failed to load payment: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error fetching payment: $e');
    }
  }

  // Get user's payment history
  static Future<List<PaymentModel>> getUserPayments(String userId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/payments/user/$userId'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body)['data'];
        return data.map((json) => PaymentModel.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load user payments: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error fetching user payments: $e');
    }
  }

  // Get payment status
  static Future<PaymentModel> getPaymentStatus(String paymentId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/payments/$paymentId/status'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body)['data'];
        return PaymentModel.fromJson(data);
      } else {
        throw Exception('Failed to get payment status: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error getting payment status: $e');
    }
  }

  // Refund payment
  static Future<PaymentModel> refundPayment(String paymentId, {double? amount}) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/payments/$paymentId/refund'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: json.encode({
          'amount': amount,
        }),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body)['data'];
        return PaymentModel.fromJson(data);
      } else {
        throw Exception('Failed to refund payment: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error refunding payment: $e');
    }
  }

  // Save payment card
  static Future<PaymentCardModel> savePaymentCard({
    required String userId,
    required String cardNumber,
    required String cardholderName,
    required int expiryMonth,
    required int expiryYear,
    required String cvv,
    bool isDefault = false,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/payments/cards'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: json.encode({
          'userId': userId,
          'cardNumber': cardNumber,
          'cardholderName': cardholderName,
          'expiryMonth': expiryMonth,
          'expiryYear': expiryYear,
          'cvv': cvv,
          'isDefault': isDefault,
        }),
      );

      if (response.statusCode == 201) {
        final data = json.decode(response.body)['data'];
        return PaymentCardModel.fromJson(data);
      } else {
        throw Exception('Failed to save payment card: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error saving payment card: $e');
    }
  }

  // Get user's saved payment cards
  static Future<List<PaymentCardModel>> getUserPaymentCards(String userId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/payments/cards/user/$userId'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body)['data'];
        return data.map((json) => PaymentCardModel.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load payment cards: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error fetching payment cards: $e');
    }
  }

  // Delete payment card
  static Future<void> deletePaymentCard(String cardId) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/payments/cards/$cardId'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to delete payment card: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error deleting payment card: $e');
    }
  }

  // Set default payment card
  static Future<PaymentCardModel> setDefaultPaymentCard(String cardId) async {
    try {
      final response = await http.patch(
        Uri.parse('$baseUrl/payments/cards/$cardId/default'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body)['data'];
        return PaymentCardModel.fromJson(data);
      } else {
        throw Exception('Failed to set default card: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error setting default card: $e');
    }
  }

  // Get payment methods available
  static Future<List<Map<String, dynamic>>> getAvailablePaymentMethods() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/payments/methods'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body)['data'];
        return data.cast<Map<String, dynamic>>();
      } else {
        throw Exception('Failed to get payment methods: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error fetching payment methods: $e');
    }
  }

  // Validate card number (Luhn algorithm)
  static bool validateCardNumber(String cardNumber) {
    if (cardNumber.isEmpty) return false;
    
    // Remove spaces and dashes
    final cleanNumber = cardNumber.replaceAll(RegExp(r'[\s-]'), '');
    
    if (cleanNumber.length < 13 || cleanNumber.length > 19) return false;
    
    int sum = 0;
    bool isEven = false;
    
    // Loop through values starting from the rightmost side
    for (int i = cleanNumber.length - 1; i >= 0; i--) {
      int digit = int.parse(cleanNumber[i]);
      
      if (isEven) {
        digit *= 2;
        if (digit > 9) {
          digit -= 9;
        }
      }
      
      sum += digit;
      isEven = !isEven;
    }
    
    return sum % 10 == 0;
  }

  // Get card type from number
  static String getCardType(String cardNumber) {
    final cleanNumber = cardNumber.replaceAll(RegExp(r'[\s-]'), '');
    
    if (cleanNumber.startsWith('4')) {
      return 'Visa';
    } else if (cleanNumber.startsWith('5')) {
      return 'Mastercard';
    } else if (cleanNumber.startsWith('3')) {
      return 'American Express';
    } else if (cleanNumber.startsWith('6')) {
      return 'Discover';
    } else {
      return 'Unknown';
    }
  }

  // Validate expiry date
  static bool validateExpiryDate(int month, int year) {
    final now = DateTime.now();
    final currentYear = now.year;
    final currentMonth = now.month;
    
    if (year < currentYear) return false;
    if (year == currentYear && month < currentMonth) return false;
    if (month < 1 || month > 12) return false;
    
    return true;
  }

  // Validate CVV
  static bool validateCVV(String cvv, String cardType) {
    if (cvv.isEmpty) return false;
    
    final cvvLength = cvv.length;
    if (cardType == 'American Express') {
      return cvvLength == 4;
    } else {
      return cvvLength == 3;
    }
  }
}

File: apps/web/src/services/paymentService.js
import api from './api';

const paymentService = {
  // Create a new payment
  createPayment: async (paymentData) => {
    const response = await api.post('/payments/create-payment', paymentData);
    return response.data;
  },

  // Confirm payment status
  confirmPayment: async (paymentId) => {
    const response = await api.post(`/payments/confirm-payment/${paymentId}`);
    return response.data;
  },

  // Get payment status
  getPaymentStatus: async (paymentId) => {
    const response = await api.get(`/payments/${paymentId}/status`);
    return response.data;
  },

  // Process refund
  processRefund: async (bookingId, refundData) => {
    const response = await api.post(`/payments/refund/${bookingId}`, refundData);
    return response.data;
  },

  // Save payment method
  savePaymentMethod: async (paymentMethodData) => {
    const response = await api.post('/payments/save-method', paymentMethodData);
    return response.data;
  },

  // Get user's saved payment methods
  getUserPaymentMethods: async () => {
    const response = await api.get('/payments/methods');
    return response.data;
  },

  // Moyasar SDK initialization
  initializeMoyasar: () => {
    const script = document.createElement('script');
    script.src = 'https://cdn.moyasar.com/mpf/1.5.3/moyasar.js';
    script.async = true;
    document.head.appendChild(script);
    
    return new Promise((resolve, reject) => {
      script.onload = () => {
        if (window.Moyasar) {
          window.Moyasar.init({
            element: '.moyasar-form',
            amount: 0, // Will be set dynamically
            currency: 'SAR',
            description: 'LUDUS Activity Booking',
            publishable_api_key: process.env.REACT_APP_MOYASAR_PUBLISHABLE_KEY,
            callback_url: `${window.location.origin}/payment/callback`,
            methods: ['creditcard', 'applepay', 'stcpay', 'aman'],
            on_initiating: () => {
              console.log('Payment initiated');
            },
            on_completed: (payment) => {
              console.log('Payment completed:', payment);
            },
            on_error: (error) => {
              console.error('Payment error:', error);
            }
          });
          resolve(window.Moyasar);
        } else {
          reject(new Error('Moyasar SDK failed to load'));
        }
      };
      script.onerror = () => reject(new Error('Failed to load Moyasar SDK'));
    });
  },

  // Create Moyasar payment form
  createPaymentForm: (containerId, options = {}) => {
    if (!window.Moyasar) {
      throw new Error('Moyasar SDK not loaded. Call initializeMoyasar() first.');
    }

    const defaultOptions = {
      element: `#${containerId}`,
      amount: options.amount || 0,
      currency: 'SAR',
      description: options.description || 'LUDUS Activity Booking',
      publishable_api_key: process.env.REACT_APP_MOYASAR_PUBLISHABLE_KEY,
      callback_url: options.callbackUrl || `${window.location.origin}/payment/callback`,
      methods: options.methods || ['creditcard', 'applepay', 'stcpay', 'aman'],
      metadata: options.metadata || {},
      on_initiating: options.onInitiating || (() => {}),
      on_completed: options.onCompleted || (() => {}),
      on_error: options.onError || (() => {})
    };

    return window.Moyasar.init(defaultOptions);
  },

  // Validate card data
  validateCardData: (cardData) => {
    const errors = {};

    if (!cardData.number || cardData.number.length < 16) {
      errors.number = 'Card number must be at least 16 digits';
    }

    if (!cardData.name || cardData.name.trim().length < 2) {
      errors.name = 'Cardholder name is required';
    }

    if (!cardData.month || cardData.month < 1 || cardData.month > 12) {
      errors.month = 'Valid expiry month is required';
    }

    if (!cardData.year || cardData.year < new Date().getFullYear()) {
      errors.year = 'Valid expiry year is required';
    }

    if (!cardData.cvc || cardData.cvc.length < 3) {
      errors.cvc = 'Valid CVC is required';
    }

    return {
      isValid: Object.keys(errors).length === 0,
      errors
    };
  },

  // Format card number for display
  formatCardNumber: (number) => {
    const cleaned = number.replace(/\D/g, '');
    const formatted = cleaned.replace(/(.{4})/g, '$1 ').trim();
    return formatted;
  },

  // Get card type from number
  getCardType: (number) => {
    const cleaned = number.replace(/\D/g, '');
    
    if (/^4/.test(cleaned)) return 'visa';
    if (/^5[1-5]/.test(cleaned)) return 'mastercard';
    if (/^3[47]/.test(cleaned)) return 'amex';
    if (/^6/.test(cleaned)) return 'discover';
    if (/^9/.test(cleaned)) return 'mada'; // Saudi MADA cards
    
    return 'unknown';
  },

  // Format currency for Saudi Riyals
  formatCurrency: (amount) => {
    return new Intl.NumberFormat('ar-SA', {
      style: 'currency',
      currency: 'SAR',
      minimumFractionDigits: 2
    }).format(amount);
  },

  // Convert SAR to halalas (Moyasar uses halalas)
  toHalalas: (amount) => {
    return Math.round(amount * 100);
  },

  // Convert halalas to SAR
  fromHalalas: (halalas) => {
    return halalas / 100;
  },

  // Get payment history
  getPaymentHistory: async (params = {}) => {
    const queryParams = new URLSearchParams(params);
    const response = await api.get(`/payments/history?${queryParams}`);
    return response.data;
  },

  // Get payment analytics
  getPaymentAnalytics: async (params = {}) => {
    const queryParams = new URLSearchParams(params);
    const response = await api.get(`/payments/analytics?${queryParams}`);
    return response.data;
  },

  // Get payment methods configuration
  getPaymentMethodsConfig: async () => {
    const response = await api.get('/payments/config');
    return response.data;
  },

  // Delete payment method
  deletePaymentMethod: async (methodId) => {
    const response = await api.delete(`/payments/methods/${methodId}`);
    return response.data;
  },

  // Set default payment method
  setDefaultPaymentMethod: async (methodId) => {
    const response = await api.put(`/payments/methods/${methodId}/default`);
    return response.data;
  },

  // Enhanced card validation
  validateCardDataEnhanced: (cardData) => {
    const errors = {};

    // Card number validation
    if (!cardData.number || cardData.number.length < 16) {
      errors.number = 'Card number must be at least 16 digits';
    } else {
      // Luhn algorithm validation
      if (!paymentService.validateLuhn(cardData.number)) {
        errors.number = 'Invalid card number';
      }
    }

    // Cardholder name validation
    if (!cardData.name || cardData.name.trim().length < 2) {
      errors.name = 'Cardholder name is required';
    } else if (cardData.name.trim().length > 100) {
      errors.name = 'Cardholder name is too long';
    }

    // Expiry month validation
    if (!cardData.month || cardData.month < 1 || cardData.month > 12) {
      errors.month = 'Valid expiry month is required';
    }

    // Expiry year validation
    const currentYear = new Date().getFullYear();
    if (!cardData.year || cardData.year < currentYear) {
      errors.year = 'Valid expiry year is required';
    } else if (cardData.year > currentYear + 20) {
      errors.year = 'Expiry year is too far in the future';
    }

    // CVC validation
    if (!cardData.cvc || cardData.cvc.length < 3) {
      errors.cvc = 'Valid CVC is required';
    } else if (cardData.cvc.length > 4) {
      errors.cvc = 'CVC is too long';
    }

    return {
      isValid: Object.keys(errors).length === 0,
      errors
    };
  },

  // Luhn algorithm validation
  validateLuhn: (cardNumber) => {
    const cleaned = cardNumber.replace(/\D/g, '');
    let sum = 0;
    let isEven = false;

    for (let i = cleaned.length - 1; i >= 0; i--) {
      let digit = parseInt(cleaned.charAt(i));

      if (isEven) {
        digit *= 2;
        if (digit > 9) {
          digit -= 9;
        }
      }

      sum += digit;
      isEven = !isEven;
    }

    return sum % 10 === 0;
  },

  // Get card issuer information
  getCardIssuer: (cardNumber) => {
    const cleaned = cardNumber.replace(/\D/g, '');
    
    // Visa
    if (/^4/.test(cleaned)) {
      return {
        type: 'visa',
        name: 'Visa',
        icon: '💳',
        color: '#1A1F71'
      };
    }
    
    // Mastercard
    if (/^5[1-5]/.test(cleaned)) {
      return {
        type: 'mastercard',
        name: 'Mastercard',
        icon: '💳',
        color: '#EB001B'
      };
    }
    
    // American Express
    if (/^3[47]/.test(cleaned)) {
      return {
        type: 'amex',
        name: 'American Express',
        icon: '💳',
        color: '#006FCF'
      };
    }
    
    // MADA (Saudi)
    if (/^9/.test(cleaned)) {
      return {
        type: 'mada',
        name: 'MADA',
        icon: '🏦',
        color: '#00A651'
      };
    }
    
    return {
      type: 'unknown',
      name: 'Unknown',
      icon: '💳',
      color: '#6B7280'
    };
  },

  // Format card number with spaces
  formatCardNumberWithSpaces: (number) => {
    const cleaned = number.replace(/\D/g, '');
    const formatted = cleaned.replace(/(.{4})/g, '$1 ').trim();
    return formatted;
  },

  // Mask card number for display
  maskCardNumber: (number, visibleDigits = 4) => {
    const cleaned = number.replace(/\D/g, '');
    const lastDigits = cleaned.slice(-visibleDigits);
    const masked = '*'.repeat(cleaned.length - visibleDigits);
    return masked + lastDigits;
  },

  // Calculate payment fees
  calculateFees: (amount, method = 'credit_card') => {
    const feeRates = {
      credit_card: 0.029, // 2.9%
      mada: 0.015, // 1.5%
      apple_pay: 0.029, // 2.9%
      stc_pay: 0.02, // 2%
      sadad: 0.01 // 1%
    };

    const rate = feeRates[method] || 0.029;
    const fee = amount * rate;
    const netAmount = amount - fee;

    return {
      amount,
      fee: Math.round(fee * 100) / 100,
      netAmount: Math.round(netAmount * 100) / 100,
      rate: rate * 100
    };
  },

  // Get payment status color
  getPaymentStatusColor: (status) => {
    const colors = {
      pending: '#F59E0B',
      processing: '#3B82F6',
      completed: '#10B981',
      failed: '#EF4444',
      cancelled: '#6B7280',
      refunded: '#8B5CF6'
    };
    return colors[status] || '#6B7280';
  },

  // Get payment status text
  getPaymentStatusText: (status) => {
    const texts = {
      pending: 'Pending',
      processing: 'Processing',
      completed: 'Completed',
      failed: 'Failed',
      cancelled: 'Cancelled',
      refunded: 'Refunded'
    };
    return texts[status] || 'Unknown';
  }
};

export { paymentService };